code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/LoadBalancerServiceUpdatePostListener.java

```diff
@@ -103,8 +109,18 @@ private void updateServicePorts(Service service, ProcessState state) {
                 port = objectManager.create(port);
             }
 
+            allocatorService.newPortAllocateForLoadBalancerUpdate(instance, toCreate);
+            
             // trigger instance/metadata update
-            instance = objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, newPortDefs);
+            List<Port> toUpdate = objectManager.find(Port.class, PORT.INSTANCE_ID, instance.getId(), PORT.REMOVED, null);
+            List<String> toUpdatePortDefs = new ArrayList<>();
+
+            for (Port port : toUpdate) {
+                PortSpec portSpec = new PortSpec(port);
+                toUpdatePortDefs.add(portSpec.toSpec());
+            }
+
+            instance = objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, toUpdatePortDefs);
```

Comment:
I'd put this call (objectManager.setFields(instance, InstanceConstants.FIELD_PORTS, toUpdatePortDefs) to corresponding allocatorDao method. So the listener calls allocator service, allocator service calls external scheduler to allocate the ip address, then Dao is called to persist the ports + update the instance.